### PR TITLE
Add Utop integration

### PR DESCRIPTION
**Pull request by tmattio**
This PR adds a code evaluation feature that sends selected expressions to the REPL when pressing `Shift+Enter` (or selecting "Evaluate Expression" in the editor menu).

The REPL is provided with Utop and is instantiated using `dune utop`, so this feature depends on both dune and utop to be available in the user's sandbox.

Closes https://github.com/ocamllabs/vscode-ocaml-platform/issues/452
Closes https://github.com/ocamllabs/vscode-ocaml-platform/issues/490

**TODO**

- [X] Fix code not being evaluated the first time when utop takes too long to be initialized
- [ ] Add popup to install utop if not available in the sandbox (most likely depends on #495 which adds Opam package installation capabilities)
- [x] Document the commands in the readme

### CODE REVIEW

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:5]: Is there a way to make this part of the instance state?

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:57]: Should be `let+ (_ : ..)` with a proper type annotation.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:45]: Base should have a function for suffix checking. Something like `String.is_suffix` I think.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:77]: No need for the trailing `else`.

File: src/terminal_sandbox.ml
Comment (rgrinberg) [inline src/terminal_sandbox.ml:97]: Should use `String.is_suffix` as well.

File: src/terminal_sandbox.ml
Comment (rgrinberg) [inline src/terminal_sandbox.ml:76]: Use a pattern match here instead of `Option.value`. It's not obvious that whatever is generating `~default` won't fail, so let's just be safe and only call it when needed.

File: src/vscode_ocaml_platform.ml
Comment (tmattio) [inline src/vscode_ocaml_platform.ml:17]: This should probably use executeCommand instead

File: package.json
Comment (rgrinberg) [inline package.json:151]: I think this toplevel only works for OCaml unfortunately.

File: vscode/vscode/vscode.mli
Comment (tmattio) [inline vscode/vscode/vscode.mli:307]: Should be possible to use `include module type of ... with type t :=`

**Comment by rgrinberg** [comment]
Looks good. I don't find `$ dune utop` too useful at its current state, but its nice to have this for when we do improve `$ dune utop`.

**Comment by tmattio** [comment]
>  I don't find $ dune utop too useful at its current state, but its nice to have this for when we do improve $ dune utop.

If users want to evaluate code that depends on the current library modules, that wouldn't work without `dune utop`, right? Although I probably need to add `--implicit-bindings` for it to work.

**Comment by rgrinberg** [comment]
> If users want to evaluate code that depends on the current library modules, that wouldn't work without dune utop, right? Although I probably need to add --implicit-bindings for it to work.

Yeah, we need to use dune somehow. The issue for me is that it's only possible to access the public api's via `$ dune utop`. It would be much better to have something like `$ dune utop bar/foo.ml` that will allow us to peek into the internals of `Foo`.

File: package.json
Comment (tmattio) [reply package.json:151]: That's for the destruct command, I just removed a space

File: src/repl.ml
Comment (tmattio) [reply src/repl.ml]: Good idea, done

**Comment by tmattio** [comment]
This is ready I think. I'm not too happy with the number of workarounds in there, so it would be nice to revisit the implementation once we have the toolchain ready (#463) and Dune RPC. In the meantime, the current workflow is the following:

- If `dune build` passes and the current sandbox has `utop` installed, the REPL will be `dune utop`
- If `dune build` fails and the current sandbox has `utop` installed, the REPL will be `utop` 
- Else, the REPL will be `ocaml`

File: src/terminal_sandbox.ml
Comment (rgrinberg) [inline src/terminal_sandbox.ml:98]: `not (String.is_suffix text ~suffix:"\n")`

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:148]: `else` not needed.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:14]: `Cmd.which` already exists

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:52]: What would be simpler than all of this is to just ask the user what command to use to launch the repl.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:113]: Do we want to at least log the error?

File: src/repl.ml
Comment (tmattio) [reply src/repl.ml:52]: We would still have to detect if utop is available and handle the potential errors. I'm not sure asking the user would simplify the code that much

Comment (rgrinberg) [reply src/repl.ml:42]: Sorry, I meant simplify from the point of the user. Currently, this functionality doesn't really work if the user wants to use `down`, `rtop`, or just pass some flags to `dune utop`. Also, it requires remembering a few implementation details about the project that aren't too intuitive.  If we let the user pick how they launch the toplevel we can sidestep all of this.

Comment (tmattio) [reply src/repl.ml:42]: I see, maybe we could have a setting for the command used to run the repl? This way users would have a good default, and they could change their repl if needed.

I think most users would be happy with utop, as it's the recommended repl for OCaml, so I wouldn't want to interrupt the workflow with a popup for them.

Comment (tmattio) [reply src/repl.ml]: I added settings to set the REPL path and arguments. Here are the new settings:

| Name                               | Description                                                                                             | Default |
| ---------------------------------- | ------------------------------------------------------------------------------------------------------- | ------- |
| `ocaml.repl.path.linux`            | The path of the REPL that the extension uses on Linux                                                   | `null`  |
| `ocaml.repl.path.osx`              | The path of the REPL that the extension uses on macOS                                                   | `null`  |
| `ocaml.repl.path.windows`          | The path of the REPL that the extension uses on Windows                                                 | `null`  |
| `ocaml.repl.args.linux`            | The REPL arguments that the extension uses on Linux                                                     | `null`  |
| `ocaml.repl.args.osx`              | The REPL arguments that the extension uses on macOS                                                     | `null`  |
| `ocaml.repl.args.windows`          | The REPL arguments that the extension uses on Window                                                    | `null`  |

Comment (rgrinberg) [reply src/repl.ml]: Why is it per platform? I get that it might be useful in some cases, but most users only use a single platform.

Comment (tmattio) [reply src/repl.ml]: Agree it's not very useful, I removed them in favor of `ocaml.repl.path` and `ocaml.repl.args`

**Comment by tmattio** [comment]
@rgrinberg This is ready to merge if the addition of the settings looks good to you.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:3]: `Repl_path`.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:19]: `Repl_args`.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:45]: This check isn't quite right for `$ dune utop` since we don't use the binary. In any case, I suggest that we use `Bin.which "utop"` to make a check like this (as we do elsewhere).

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:136]: This binding occurs in both `then` and `else` clauses. Hence it can exist outside the `if` altogether.

File: vscode/vscode/vscode.mli
Comment (mnxn) [inline vscode/vscode/vscode.mli:266]: These lines can be combined as 

```ocaml
include module type of Range with type t = private Range.t
```

File: src/bindings/node/node.ml
Comment (mnxn) [inline src/bindings/node/node.ml:10]: Use `include Class.Make ()` instead.

**Review by mnxn — CHANGES_REQUESTED** [review]

File: src/repl.ml
Comment (tmattio) [reply src/repl.ml:45]: `Cmd.check` is global so doesn't do exactly what we want here. Let's say the user has utop in its path (globally installed, or in the global switch), `Cmd.check` will return `Ok`, but `dune utop` and `opam exec -- utop` would fail.

Comment (rgrinberg) [reply src/repl.ml:45]: Yes, but you can fix this by passing the PATH of the sandbox to which.

Comment (tmattio) [reply src/repl.ml:45]: True, although we'd need to get the env differently for Esy, and we'd need to run a command (i.e. `opam env` or the esy equivalent) and parse them to access the env of the sandbox. The benefit of `Cmd.which` is that we can save a command execution, which won't be the case in this specific case.

Comment (rgrinberg) [reply src/repl.ml:45]: > although we'd need to get the env differently for Esy, and we'd need to run a command (i.e. opam env or the esy equivalent) and parse them to access the env of the sandbox

We had code that did this before. I wonder where it went 🤔 In any case it would be useful to have it.

> The benefit of Cmd.which is that we can save a command execution, which won't be the case in this specific case.

My goal isn't to shave off a command invocation, but to stop relying on the command lines of these tools and treat them all consistently. However, I just checked and we already rely on `--version` for lsp, so I suppose we can just leave this be and get rid of this stuff another time.

Comment (tmattio) [reply src/repl.ml:45]: > We had code that did this before. I wonder where it went

I did something to use a custom `env` for `Cmd.check` and `Cmd.output` in #463, and actually ported these bits in this PR, but I don't recall seeing code that calls `opam env`.

**Comment by tmattio** [comment]
@rgrinberg @mnxn let me know if there's anything else you'd like me to change

**Review by mnxn — APPROVED** [review]

**Review by rgrinberg — APPROVED** [review]

**Comment by d4hines** [comment]
I would _love_ to try this one out! Any chance we can get a release (or a beta/insiders release?). Thanks!

**Comment by ulugbekna** [comment]
@d4hines in case you don't want to wait for the release, you can try the new extension by cloning the repo, installing the dependencies using opam, and running `make install`.

**Comment by smorimoto** [comment]
Alternatively, you can select the latest master build here to download the artifact and install it. This is probably the easiest way.
https://github.com/ocamllabs/vscode-ocaml-platform/actions/workflows/main.yml?query=branch%3Amaster

```
wget <URL>
unzip <PATH>
code --install-extension <FILE>
```

**Comment by d4hines** [comment]
Thanks for the advice @ulugbekna and @smorimoto ; however, I'm having issues: any command I try to run in gives me an error like:
```
Command 'OCaml: Select a Sandbox for this Workspace' resulted in an error (command 'ocaml.select-sandbox' not found)
```

To clarify, I'm on Windows10, running the Insider's build of VS Code, using the Remote WSL extension to develop on Arch Linux WSL2, so it might be a perfect storm 😅

**Comment by d4hines** [comment]
I have no idea why, but the issue was that whenever I installed the VSIX manually or with `make install`, many node_modules ended up missing. Running `yarn` inside the `~/.vscode-server/extensions/ocaml-platform-blah-blah` folder fixed the issue.

**Pull request merged**
